src/logger/InteractiveLogger.js

```diff
@@ -0,0 +1,263 @@
+    const color = getColorForString(taskName)
+    const prefix = color.fg(`${taskName} `)
+
+    this.moveCursorToAboveTasksSection()
+
+    /** @type {Task} */
+    const task = {
+      prefix: taskName,
+      coloredPrefix: prefix,
+      lastLogLine: '',
+      status: 'waiting',
+      startedAt: start,
+    }
+
+    this.tasks.push(task)
+    this.printTasks()
+
+    let isDone = false
+    const assertNotDone = () => {
+      if (isDone) {
+        throw new Error('This task has already been completed')
+      }
+    }
+
+    /** @type {string[]} */
+    const bufferedLogLines = []
+
+    const log = (/** @type {TaskStatus} */ status, /** @type {string[]} */ ...args) => {
+      assertNotDone()
+
+      // if the status changes, we need to reprint the tasks immediately. if
+      // not, we can just wait for our interval loop to do it
+      const didStatusChange = task.status !== status
+      if (didStatusChange) {
+        this.moveCursorToAboveTasksSection()
+        task.status = status
+      }
+
+      const message = args.join(' ')
+      bufferedLogLines.push(message)
+      const lastLine = lastNonEmptyLineIfPossible(message)
+      if (lastLine) {
+        task.lastLogLine = lastLine
+      }
+
+      if (didStatusChange) {
+        this.printTasks()
+      }
+    }
+
+    const complete = (/** @type {TaskStatus} */ status, /** @type {string} */ message) => {
+      assertNotDone()
+      this.moveCursorToAboveTasksSection()
+      this.tty.clearScreenDown()
+      task.status = status
+      const lastLine = lastNonEmptyLineIfPossible(message)
+      if (lastLine) {
+        task.lastLogLine = lastLine
+      }
+
+      bufferedLogLines.push(message)
+      this.tty.write(color.fg(':: ') + color.bg().bold(` ${taskName} `) + color.fg(' ::') + '\n')
+      for (const line of bufferedLogLines) {
+        this.tty.write(line + '\n')
+      }
+      this.tty.write('\n')
+
+      isDone = true
+      this.printTasks()
+    }
+
+    return {
+      restartTimer: () => {
+        assertNotDone()
+        start = Date.now()
+        task.startedAt = start
+      },
+      log: (...args) => {
+        log('running', ...args)
+      },
+      logErr: (...args) => {
+        log('running', ...args)
+      },
+      fail: (headline, more) => {
+        complete('failed', formatFailMessage(headline, more))
+      },
+      success: (message) => {
+        complete('done', formatSuccessMessage(message, k.gray(`in ${timeSince(start)}`)))
+      },
+      info: (...args) => {
+        log('running', formatInfoMessage(...args))
+      },
+      note: (...args) => {
+        log('running', formatNoteMessage(...args))
+      },
+    }
+  }
+}
+
+/**
+ * @param {Task} a
+ * @param {Task} b
+ */
+function compareTasksForPrinting(a, b) {
+  if (a.startedAt < b.startedAt) {
+    return -1
+  }
+  if (a.startedAt > b.startedAt) {
+    return 1
+  }
+  if (a.prefix < b.prefix) {
+    return -1
+  }
+  if (a.prefix > b.prefix) {
+    return 1
+  }
```

Comment:
Idea for the future that may totally suck but I'd like to try anyway:

```
<list of completed tasks, ordered by completion time + prefix (earliest first)>
<blank line>
<list of running tasks, ordered by start time + prefix>
```

Then when tasks complete they kinda pop up into the upper section. Could be a fun juicy flourish. Also might just be confusing/disorienting.

Reply:
I tried this before but found things jumping around like that to be a little distracting / hard to follow. Keeping things stable seemed the easiest to follow